[MRG] Add fast kernel classifier/regressor

Alex7Li opened the pull request:
This pull request was originally for sklearn, but it was decided to move it to sklearn-extra https://github.com/scikit-learn/scikit-learn/pull/11694. We add the EigenPro, also known as FastKernel method.

Alex7Li commented:
Oh wow I was struggling for so long because of some mysterious error about having NaN's in a matrix that only appeared when I pushed the code here, but it just magically disappeared when I added the fastfood import (I was removing it because my code wouldn't compile in my environment with that line). Not sure why that happened, but it seems like it works now and the code is ready.

NicolasHug commented:
Thanks for the PR!

I do appreciate the code being well-documented.

Here are a few general comments after a first glance, in order of importance:

- The current tests are mostly high level, functional tests. We will also need some low level tests, e.g. individually testing the private helpers (if relevant, I haven't checked in too much details). We will also need tests for input checking [see e.g. here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py#L39) and making sure the default values behave  as expected. 
- I think it would be more appropriate to have a base class and 2 child classes, one for the regressor, one for the classifier. This is how we do things in scikit-learn. You can take inspiration e.g. at the [GBDT code](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py).
- Some of the attributes (typically Q_, V_) should not be attributes. If they're only useful at fit time, they should be destroyed when `fit` returns. Pass them along and return them from your helper methods. Same for `random_state_`.
- The doc for the estimators isn't generated.
- Please add a short description of each test, indicating what it checks, and why you would expect a given result. Also I feel like a lot of them could be factorized via parametrization.
- Please document the attributes. The public attribute should end with `_`, while the private ones should start with `_`. It seems like a lot of the current public attributes should actually be private. 
- please check your code style with flake or equivalent tool.
- The benchmark part of the user guide is redundant with the examples. It would be helpful to expand a bit the first part, and give a bit more details on the algorithm.
- Please only use lower case for variable names (e.g. `FKR_prediction` should just be `fkr_prediction` (sometimes `predictions` is enough depending on the context))
- The examples take some time to run. Maybe keep one and convert the others as benchmarks?
- CI is still failing

NicolasHug commented:
Also, for variable that aren't attributes, and if there isn't a more obvious verbose alternative, it'd be great if you followed the paper notation (e.g. E and \Lambda as the result of the SVD).

rth commented:
Thanks for this PR @Alex7Li . Looking forward to being able to use it!

I'll review in more detail once the above comments are addressed. Also please add new estimators to the [list here](https://github.com/scikit-learn-contrib/scikit-learn-extra/blob/fba8ffdf52d1e195d9261649a4df2eccfa8917db/sklearn_extra/tests/test_common.py#L10) so that `check_estimator` is run on them.

GaelVaroquaux commented:
Playing with this code in my lab, we are struggling with the number of eigenvalues to approximate the kernel: setting it to a large value throws an error on certain datasets, as a result we must set it to a moderate value (160, as in the paper), which is suboptimal for certain datasets. It ends up being a dataset-specific parameter, which is a usability concern.

Cc @pcerda

amueller commented:
@GaelVaroquaux what's the error? Can you provide code to reproduce?

GaelVaroquaux commented:
> @GaelVaroquaux what's the error?

Nan

> Can you provide code to reproduce?

Ping @pcerda

pcerda commented:
> @GaelVaroquaux what's the error?
> Nan
> Can you provide code to reproduce?
> Ping @pcerda

The problem comes in line 177 of fast_kernel.py when calculating the highest `n_components` eigenvalues:
```
S, V = sp.linalg.eigh(W, eigvals=(m - n_components, m - 1))
`
where `n_components=1000` and `m=1601`
with the following hermitian matrix W:
[https://www.dropbox.com/s/xqcxhnx1p06u76o/W.npy?dl=1](url)

if `n_components=160` I do not have this problem.

amueller commented:
@pcerda can you share the full code and data? If the data is private, can you share a synthetic or open example that reproduces the issue?

amueller commented:
It's a bit annoying that you can't see all the failing tests at once, and we want to fix that. It's not super easy. Please let us know if there's other issues.

rth commented:
@Alex7Li Please merge master in, with https://github.com/scikit-learn-contrib/scikit-learn-extra/pull/18 that should fix failing docstrings. Also please run [black](https://github.com/python/black) on the added/changed files, that will fix linter CI.

In general if you encounter any issues, or want someone to look at the latest changes don't hesitate to comment on this PR. Do you have any comments about https://github.com/scikit-learn-contrib/scikit-learn-extra/pull/13#issuecomment-503138825 ?

Alex7Li commented:
Thanks a lot for the kind words and help with solving the documentation issue. I have talked with Siyuan Ma and worked on the code to resolve the issue in comment 13. We determined that the issue was related to precision errors when calculating eigenvectors - the matrix W should be positive definite but when calculating the eigenvalues, numerical error can give negative eigenvalues and result in a LinAlgError as was seen (it took a long time to fix because even the isolated error in the comment did not show up locally, but I eventually got the code to crash by running it on the server in a different version of python than the default). I found that using float64 rather than float32 fixed the error. Since that line can be performance critical depending on input parameters, I surrounded it in a try catch - try to use the float32 solution, and if it fails, run it with float64. I don't think that the float64 solution is likely to break, considering how hard it was to get even float32 numbers to not work. If there is still concern, we can add yet another try-catch and run scipy.linalg.eig instead of eigh. 

After investigating the last error in travis ci, I do not believe there is anything I can do to fix it. It is caused by us failing on the test-common for earlier versions of sklearn. The test we fail tests this: "If the estimator does not support multi-output, then it should give a warning". Since this estimator does support multi-output, this test does not apply, and on earlier pushes, this was not a problem since we added the estimator to the list of estimators that were multi-output so the code ignored it. However, in the later versions, rather than keeping a list, the multi-output estimators were supposed to return tags in the get_tags function, and that is working properly which is why we don't fail on those. There is no way to modify the list of multi-output estimators for the earlier versions of sklearn that I know of, though, so I don't know how to get rid of the error and make it a passing build.

amueller commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,641 @@
+        it will be 4000 if there are less than 100,000 samples
+        (for training), and otherwise 10000.
+
+    kernel : string or callable, default = "gaussian"
```
Maybe calling ``gaussian`` ``rbf`` might sense to make it more consistent with sklearn? What do you mean by a radial kernel here?

Alex7Li replied:
I will change the name and try to make the description more clear

amueller replied:
Awesome, thanks!

amueller commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,641 @@
+        Ignored by other kernels.
+
+    gamma : float, default=None
+        Gamma parameter for the RBF, polynomial, exponential chi2 and
```
This is confusing to me as this should be the same as bandwidth, right?

Alex7Li replied:
I think you are right that it is redundant, I was copying Kernel Ridge's interface for extra kernels it didn't really occur to me. I will remove the parameter

amueller commented:
Hm once https://github.com/scikit-learn/scikit-learn/pull/14381 is merged we'll be able to get more helpful error messages. Though I guess not on the CI because it's not release yet?

This seems to be quite helpful according to initial experiments so I'd love to see this merged.

Alex7Li commented:
I think that it's ready to be merged right now, I explained [above ](https://github.com/scikit-learn-contrib/scikit-learn-extra/pull/13#issuecomment-508307884) why I think it's impossible to fix the CI errors. Is there something else I should do?

amueller commented:
oh, you're right. Sorry, I lost track. I didn't review this in detail but I think it's in a state where we can iterate from, which I think is the point of scikit-learn-extra.
Maybe remove the underscore in the class name, I don't think that's pep8? https://www.python.org/dev/peps/pep-0008/#class-names

amueller commented:
@NicolasHug you had looked at this before, do you want to do another round?

Alex7Li commented:
Oh hey it passed the tests cause the version changed!

rth commented:
> Oh hey it passed the tests cause the version changed!

Great! Yeah we dropped support for scikit-learn <0.20 recently for this repo.

rth commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,680 @@
+    """
+
+    @abstractmethod
+    def __init__(
```
I'm not sure that `__init__` should be abstract method, you are calling it inherited classes. Making a regular method would probably be better. And then you can remove the `ABC` class.

Alex7Li replied:
Done

rth commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,680 @@
+    batch_size : int, default = 'auto'
+        Mini-batch size for gradient descent.
+
+    n_epoch : int, default = 1
```
The default is 2.

Alex7Li replied:
Right, fixed

rth commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,680 @@
+        The number of subsamples used for estimating the largest
+        n_component eigenvalues and eigenvectors. When it is set to 'auto',
+        it will be 4000 if there are less than 100,000 samples
+        (for training), and otherwise 10000.
```
```suggestion
        (for training), and otherwise 12000.
```

Alex7Li replied:
Fixed

rth commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,680 @@
+
+    kernel : string or callable, default = "rbf"
+        Kernel mapping used internally. Strings can be anything supported
+        by sklearn's library, however, there is special support for the
```
```suggestion
        by scikit-learn, however, there is special support for the
```

Alex7Li replied:
Changed.

rth commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,680 @@
+        class_matrix = np.zeros((Y.shape[0], self.classes_.shape[0]))
+
+        for ind, label in enumerate(Y):
+            class_matrix[ind][loc[label]] = 1
```
```suggestion
            class_matrix[ind, loc[label]] = 1
```

Alex7Li replied:
Replaced with your suggestion

rth commented on sklearn_extra/tests/test_fast_kernel.py:
```diff
@@ -0,0 +1,251 @@
+
+import pytest
+
+np.random.seed(1)
```
We should not set the random seed globally as it would affect all other test.

Use `rng = np.random.RandomState(1)` when needed instead.

Alex7Li replied:
I just removed the line since I don't even use np.random

rth commented on sklearn_extra/tests/test_fast_kernel.py:
```diff
@@ -0,0 +1,251 @@
+    """
+    X, y = data
+    prediction = estimator.fit(X, y).predict(X)
+    assert_array_almost_equal(abs(prediction / y) / 2.0, 0.5, decimal=2)
```
What is `abs(prediction / y) / 2.0`?

Better to use,
```py
assert_allclose(prediction, y, rtol=1e-3)
```
(or some other appropriate tolerance)

Alex7Li replied:
Replaced using 5e-3 in all instances

rth commented on sklearn_extra/tests/test_fast_kernel.py:
```diff
@@ -0,0 +1,251 @@
+        .fit(X, y)
+        .predict(X)
+    )
+    assert_array_almost_equal(abs(fkr_prediction / y) / 2.0, 0.5, decimal=2)
```
Same as above

rth commented on sklearn_extra/tests/test_fast_kernel.py:
```diff
@@ -0,0 +1,251 @@
+    """
+    X, y = data
+    prediction = estimator.fit(X, y).predict(X)
+    assert_array_almost_equal(prediction, y)
```
Better to use `assert_allclose`

rth commented on sklearn_extra/tests/test_fast_kernel.py:
```diff
@@ -0,0 +1,251 @@
+        .fit(X, y)
+        .predict(X)
+    )
+    assert_array_almost_equal(fkc_prediction, y)
```
idem

rth commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,680 @@
+    batch_size : int, default = 'auto'
+        Mini-batch size for gradient descent.
+
+    n_epoch : int, default = 1
```
```
    n_epoch : int, default = 2
```

Alex7Li commented on sklearn_extra/fast_kernel.py:
```diff
@@ -0,0 +1,680 @@
+
+                self.coef_[batch_inds] = (
+                    self.coef_[batch_inds] - step * gradient
+                )
```
Replaced with your suggestion

Alex7Li commented:
Still working on the second post (I want to rerun all the examples on our server), but I added all the changes from the first post.  I changed the class names to EigenProRegression/Classification, and then changed the main file name from fast_kernel.py to eigenpro.py as well, just to get rid of that alias. I'm not sure why it should be moved to a kernel_methods folder exactly, how does that make it easier to import?

Alex7Li commented:
Ok, I have finished updating the code

amueller commented:
I'm +1 for merging as well.
@Alex7Li it would be awesome if you could stick around a bit to help with the implementation. I think this looks very promising but it's likely that we'll run into issues (because that always happens ;).

Alex7Li commented:
Thanks Roman for explaining, I moved and finished the merge. I also tried to fix the memory error issue, I think the estimator was just choosing a batch size that was quite large.
I will stick around after this is finished, though I'm not sure exactly what that entails. Browsing through the issues and seeing if there is anything related to this code is what I'm imagining now.
Also I'm having a bit of trouble passing the checks, it seems like it's failing while trying to download packages. I don't think anything I changed should effect that, though?

rth commented:
The problem is with conda failing to downgrade from 4.7 to 4.6. I can't push to this PR to fix it (as you are using your master branch of your fork instead of a separate branch). CI was green before the last minor of commits, and tests pass locally for me. So let's just merge this. I'll make a follow up PR for fix Travis.

Thanks again @Alex7Li !

> I will stick around after this is finished, though I'm not sure exactly what that entails. Browsing through the issues and seeing if there is anything related to this code is what I'm imagining now.

Yeah, mostly that. Also if you think of anything which in your opinion could be improved (with this code or other code in scikit-learn-extra) don't hesitate to open issues.

rth commented:
@Alex7Li Also note that because you were using the master branch of your fork, if you want to sync it with the upstream master, you would need to `git reset --hard` to some commit deep in the history before your changes, and then rebase on top of `upstream/master` to get the merged changes.

For next PRs it's better to create a new branch from master..

amueller commented:
I'm a bit surprised by the current behavior, the gamma parameter is completely ignored, despite the documentation.
